File: examples/basic_usage.py
#!/usr/bin/env python3
"""
Basic usage example for the tarzi Python library.
"""

import tarzi


def main():
    html_input = """
    <html>
        <head><title>Example Page</title></head>
        <body>
            <h1>Welcome to Tarzi</h1>
            <p>This is a <strong>test</strong> page with <a href="https://example.com">a link</a>.</p>
            <img src="image.jpg" alt="Test image">
        </body>
    <
    """

    # Converter
    converter = tarzi.Converter()
    yaml_output = converter.convert(html_input, "yaml")
    print(f"YAML output:\n{yaml_output}\n")

    # WebFetcher with different modes
    fetcher = tarzi.WebFetcher()
    try:
        # Plain request mode
        content = fetcher.fetch("https://httpbin.org/html", "plain_request", "html")
        print(f"Plain request - Fetched content length: {len(content)}")

        # Raw fetch mode
        raw_content = fetcher.fetch("https://httpbin.org/html", "plain_request", "html")
        print(f"Raw fetch - Content length: {len(raw_content)}")
    except Exception as e:
        print(f"Fetch failed: {e}")

    # SearchEngine
    search_engine = tarzi.SearchEngine()
    try:
        # Note: API keys are now configured per provider in the configuration
        # See the configuration example below for how to set up API keys

        results = search_engine.search("machine learning", 2)
        print(f"Found {len(results)} search results:")
        for i, result in enumerate(results):
            print(f"  {i+1}. {result.title} ({result.url})")
            print(f"     {result.snippet}")
    except Exception as e:
        print(f"Search failed: {e}")

    # Configuration-based usage
    try:
        # Create config from string
        config_str = """
[fetcher]
timeout = 30
format = "markdown"
web_driver = "chromedriver"

[search]
engine = "bing"
"""
        config = tarzi.Config.from_str(config_str)
        print("Created config from string successfully")

        # Use config with fetcher
        tarzi.WebFetcher.from_config(config)
        print("Created fetcher from config successfully")

        # Use config with search engine
        tarzi.SearchEngine.from_config(config)
        print("Created search engine from config successfully")

    except Exception as e:
        print(f"Configuration usage failed: {e}")


if __name__ == "__main__":
    main()

File: examples/basic_usage.rs
use tarzi::{config::Config, FetchMode, Format, Result, SearchEngine, WebFetcher};

#[tokio::main]
async fn main() -> Result<()> {
    // Initialize tracing
    tracing_subscriber::fmt::init();

    println!("=== Tarzi Modular Example ===\n");

    // Load configuration and configure Chrome driver
    let mut config = Config::load().unwrap_or_default();
    config.fetcher.web_driver = "chromedriver".to_string();

    // Example 1: Using the fetcher module directly
    println!("1. Fetching content with different modes:");

    let mut fetcher = WebFetcher::from_config(&config);
    let test_url = tarzi::constants::HTTPBIN_HTML_URL;

    // Plain request mode
    match fetcher
        .fetch(test_url, FetchMode::PlainRequest, Format::Html)
        .await
    {
        Ok(content) => println!("   Plain request: {} characters", content.len()),
        Err(e) => println!("   Plain request failed: {e}"),
    }

    // Browser headless mode
    match fetcher
        .fetch(test_url, FetchMode::BrowserHeadless, Format::Markdown)
        .await
    {
        Ok(content) => println!("   Browser headless: {} characters", content.len()),
        Err(e) => println!("   Browser headless failed: {e}"),
    }

    println!();

    // Example 2: Using the search module
    println!("2. Searching and fetching content:");

    let mut search_engine = SearchEngine::from_config(&config);
    let query = "agentic AI";

    // Search for results
    match search_engine.search(query, 3).await {
        Ok(results) => {
            println!("Found {} results:", results.len());
            for (i, result) in results.iter().enumerate() {
                println!("{}. {}", i + 1, result.title);
                println!("   URL: {}", result.url);
                println!("   Snippet: {}", result.snippet);
            }
        }
        Err(e) => {
            eprintln!("Search failed: {e}");
        }
    }

    println!();

    // Example 3: Search and fetch content for each result
    println!("3. Search and fetch content for each result:");

    let results = search_engine
        .search_with_content(query, 3, FetchMode::BrowserHeadless, Format::Markdown)
        .await?;

    println!("Found {} results with content:", results.len());
    for (i, (result, content)) in results.iter().enumerate() {
        println!("{}. {}", i + 1, result.title);
        println!("   URL: {}", result.url);
        println!("   Content length: {} characters", content.len());
    }

    println!();
    println!("=== Example completed ===");

    Ok(())
}
